File: scripts/shared/mock_importer.py
"""Importador sintético usado somente em desenvolvimento e testes."""

from collections.abc import Iterator

from importer import (
    CanonicalRecord,
    ImportInspection,
    RawRecord,
    SourceFile,
    ValidationResult,
)


class MockImporter:
    name = "mock-json-importer"
    version = "1.0.0"

    def supports(self, source_file: SourceFile) -> bool:
        return source_file.media_type == "application/json"

    def inspect(self, source_file: SourceFile) -> ImportInspection:
        return ImportInspection(self.supports(source_file), 1)

    def parse(self, source_file: SourceFile) -> Iterator[RawRecord]:
        del source_file
        yield RawRecord(
            "measurement",
            {
                "metric_code": "weight",
                "value": "78.4",
                "unit_code": "kg",
                "measured_at": "2026-06-25T10:00:00Z",
            },
        )

    def normalize(self, record: RawRecord) -> CanonicalRecord:
        payload = record.payload
        return CanonicalRecord(
            metric_code=str(payload["metric_code"]),
            value=str(payload["value"]),
            unit_code=str(payload["unit_code"]),
            measured_at=str(payload["measured_at"]),
            raw_payload=payload,
        )

    def validate(self, record: CanonicalRecord) -> ValidationResult:
        try:
            float(record.value)
        except ValueError:
            return ValidationResult(False, ("value must be numeric",))
        return ValidationResult(True)

File: packages/domain/src/mock-importer.ts
import type {
  CanonicalMeasurement,
  HealthDataImporter,
  ImportInspection,
  ImportSourceFile,
  RawRecord,
  ValidationResult,
} from "./index";

export class MockImporter implements HealthDataImporter {
  readonly name = "mock-json-importer";
  readonly version = "1.0.0";

  supports(sourceFile: ImportSourceFile): boolean {
    return sourceFile.mediaType === "application/json";
  }

  async inspect(sourceFile: ImportSourceFile): Promise<ImportInspection> {
    return { supported: this.supports(sourceFile), estimatedRecords: 1 };
  }

  async *parse(_sourceFile: ImportSourceFile): AsyncIterable<RawRecord> {
    yield {
      type: "measurement",
      payload: {
        metricCode: "weight",
        value: "78.4",
        unitCode: "kg",
        measuredAt: "2026-06-25T10:00:00Z",
      },
    };
  }

  normalize(record: RawRecord): CanonicalMeasurement {
    const payload = record.payload as Record<string, string>;
    return {
      metricCode: payload.metricCode ?? "unknown",
      value: payload.value ?? "0",
      originalValue: payload.value ?? "0",
      unitCode: payload.unitCode ?? "unknown",
      originalUnitCode: payload.unitCode ?? "unknown",
      measuredAt: new Date(payload.measuredAt ?? 0),
      sourceType: "file",
      rawPayload: record.payload,
    };
  }

  validate(record: CanonicalMeasurement): ValidationResult {
    const valid =
      record.metricCode !== "unknown" &&
      Number.isFinite(Number(record.value)) &&
      !Number.isNaN(record.measuredAt.valueOf());
    return { valid, errors: valid ? [] : ["Registro canônico inválido"] };
  }
}
